File: frontends/plotly/src/metrics.ts
export interface Sample {
  seq: number;
  generation: number;
  client_time_ms: number;
  update_ms: number;
  receive_age_ms: number | null;
  skipped: number;
  conversion_ms: number;
  draw_ms: number;
  update_complete_ms: number;
}

export class MetricsSink {
  private samples: Sample[] = [];
  private active: Promise<void> | undefined;
  error: string | null = null;
  dropped = 0;

  constructor(private url: string, private mode: string, private runId: string,
    readonly metadata: Record<string, unknown>) {}

  record(sample: Sample): void {
    this.samples.push(sample);
    if (this.samples.length > 512) {
      this.dropped += this.samples.length - 512;
      this.samples.splice(0, this.samples.length - 512);
    }
  }

  async flush(includeMetadata = false): Promise<void> {
    if (this.active) return this.active;
    if (!this.samples.length && !includeMetadata) return;
    const batch = this.samples;
    this.samples = [];
    this.active = (async () => {
      try {
        const response = await fetch(`${this.url}/api/metrics`, {
          method: 'POST', headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ frontend: 'plotly', mode: this.mode, run_id: this.runId,
            samples: batch, metadata: { ...this.metadata, dropped_metric_samples: this.dropped } }),
          signal: AbortSignal.timeout(10000),
        });
        if (!response.ok) throw new Error(`Metrics upload: HTTP ${response.status}`);
        this.error = null;
      } catch (error) {
        this.error = error instanceof Error ? error.message : String(error);
        this.samples = [...batch, ...this.samples];
        if (this.samples.length > 512) {
          this.dropped += this.samples.length - 512;
          this.samples.splice(0, this.samples.length - 512);
        }
      } finally {
        this.active = undefined;
      }
    })();
    await this.active;
  }

  async close(): Promise<void> {
    await this.active;
    await this.flush(true);
  }
}

File: frontends/jfreechart/src/main/java/org/plotbench/Metrics.java
package org.plotbench;

import java.util.*;
import java.util.concurrent.*;

final class Metrics {
  final Source source;
  final String runId;
  final List<Map<String, Object>> samples = new ArrayList<>();
  final Map<String, Object> metadata = new HashMap<>();
  final ScheduledExecutorService worker =
      Executors.newSingleThreadScheduledExecutor(
          r -> {
            Thread t = new Thread(r, "plotbench-metrics");
            t.setDaemon(true);
            return t;
          });
  volatile String error;
  long lost, count, skipped;
  double lastUpdate;
  Double lastAge;

  Metrics(Source source, String runId, Map<String, Object> metadata) {
    this.source = source;
    this.runId = runId;
    this.metadata.putAll(metadata);
    worker.scheduleAtFixedRate(() -> flush(false), 1, 1, TimeUnit.SECONDS);
  }

  synchronized void set(Map<String, Object> values) {
    metadata.putAll(values);
  }

  synchronized void record(Source.Taken taken, double update, double conversion, double draw) {
    Protocol.Frame f = taken.frame();
    double now = System.currentTimeMillis();
    Double age = source.mode.equals("replay") ? null : now - f.emitted();
    Map<String, Object> sample = new HashMap<>();
    sample.put("seq", f.seq());
    sample.put("generation", f.generation());
    sample.put("client_time_ms", now);
    sample.put("update_ms", update);
    sample.put("conversion_ms", conversion);
    sample.put("draw_ms", draw);
    sample.put("receive_age_ms", age);
    sample.put("skipped", taken.skipped());
    if (samples.size() >= 4096) {
      lost++;
      error = "telemetry buffer overflow";
    } else samples.add(sample);
    count++;
    skipped += taken.skipped();
    lastUpdate = update;
    lastAge = age;
  }

  void flush(boolean last) {
    Map<String, Object> body = new HashMap<>();
    List<Map<String, Object>> batch;
    synchronized (this) {
      batch = new ArrayList<>(samples);
      samples.clear();
      Map<String, Object> meta = new HashMap<>(metadata);
      meta.put("telemetry_lost", lost);
      if (error != null) {
        meta.put("telemetry_error", error);
        if (last) meta.put("termination_reason", "error");
      }
      body.put("metadata", meta);
    }
    if (batch.isEmpty() && !last) return;
    body.put("samples", batch);
    body.put("frontend", "jfreechart");
    body.put("mode", source.mode);
    body.put("run_id", runId);
    try {
      source.post("/api/metrics", body);
    } catch (Exception ex) {
      synchronized (this) {
        lost += batch.size();
        error = ex.toString();
      }
      System.err.println("Metrics export failed: " + ex);
    }
  }

  void close() {
    worker.shutdown();
    try {
      if (!worker.awaitTermination(5, TimeUnit.SECONDS)) {
        worker.shutdownNow();
        error = "telemetry worker timeout";
      }
    } catch (InterruptedException ex) {
      Thread.currentThread().interrupt();
      error = ex.toString();
    }
    flush(true);
  }
}
